src/test/java/io/vertx/core/http/HttpBandwidthLimitingTest.java

```diff
@@ -0,0 +1,299 @@
+    HttpServer testServer = serverFactory.apply(vertx);
+    testServer.requestHandler(HANDLERS.uploadFile(sampleF));
+    startServer(testServer);
+
+    long startTime = System.nanoTime();
+    HttpClient testClient = clientFactory.apply(vertx);
+    upload(testServer, testClient, sampleF);
+    await();
+    long elapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
+
+    Assert.assertTrue( elapsedMillis >  expectedTimeMillis(Files.size(Path.of(sampleF.getAbsolutePath())), INBOUND_LIMIT));
+  }
+
+  @Test
+  public void testSendFileTrafficShapedWithSharedServers() throws InterruptedException, IOException {
+    int numEventLoops = 2; // We start a shared TCP server with 2 event-loops
+    Future<String> listenLatch = vertx.deployVerticle(() -> new AbstractVerticle() {
+      @Override
+      public void start(Promise<Void> startPromise) {
+        HttpServer testServer = serverFactory.apply(vertx);
+        testServer.requestHandler(HANDLERS.getFile(sampleF))
+                  .listen(DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST).<Void>mapEmpty().onComplete(startPromise);
+      }
+    }, new DeploymentOptions().setInstances(numEventLoops));
+
+    HttpClient testClient = clientFactory.apply(vertx);
+    CountDownLatch waitForResponse = new CountDownLatch(2);
+    AtomicLong startTime = new AtomicLong();
+    AtomicLong totalReceivedLength = new AtomicLong();
+    long expectedLength = Files.size(Path.of(sampleF.getAbsolutePath()));
+    listenLatch.onComplete(v -> {
+      startTime.set(System.nanoTime());
+      for (int i=0; i<2; i++) {
+        testClient.request(HttpMethod.GET, DEFAULT_HTTP_PORT, DEFAULT_HTTP_HOST,"/get-file")
+                  .compose(HttpClientRequest::send)
+                  .onComplete(resp -> {
+                    resp.result().bodyHandler(body -> {
+                      long receivedBytes = body.getBytes().length;
+                      totalReceivedLength.addAndGet(receivedBytes);
+                      Assert.assertEquals(expectedLength, receivedBytes);
+                      waitForResponse.countDown();
+                    });
+                  });
+      }
+    });
+    awaitLatch(waitForResponse);
+    long elapsedMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime.get());
+    Assert.assertTrue(elapsedMillis > expectedTimeMillis(totalReceivedLength.get(), OUTBOUND_LIMIT)); // because there are simultaneous 2 requests
+  }
+
+  /**
+   * The throttling takes a while to kick in so the expected time cannot be strict especially
+   * for small data sizes in these tests.
+   *
+   * @param size
+   * @param rate
+   * @return
+   */
+  private long expectedTimeMillis(long size, int rate) {
+    return (long) (TimeUnit.MILLISECONDS.convert(( size / rate), TimeUnit.SECONDS) * 0.5); // multiplied by 0.5 to be more tolerant of time pauses during CI runs
+  }
+
+  private void read(Buffer expected, HttpServer server, HttpClient client) {
```

Comment:
this code is racy (can be observed on windows CI) because the HTTP response can be received after the response handler is set and lose buffers (see https://vertx.io/docs/vertx-core/java/#_request_and_response_composition), instead should be

```java
    client.request(HttpMethod.GET, server.actualPort(), DEFAULT_HTTP_HOST,"/buffer-read")
          .compose(req -> req
            .send()
            .andThen(onSuccess(resp -> assertEquals(200, resp.statusCode())))
            .compose(HttpClientResponse::body))
          .onComplete(onSuccess(body -> {
            assertEquals(expected.getBytes().length, body.getBytes().length);
            testComplete();
          }));
```